### ISSUE TITLE
Memory leak Odata 8.2.7

### ISSUE THREAD

**Issue by Bosu123l** [OData/AspNetCoreOData#1369]
I have similar problem #873 
Issue with memory keeps growng and never released. Looks like a memory leak.
Im using im project 
Microsoft.AspNetCore.OData 8.2.7

screenshoot with private bytes from azure app service

![image](https://github.com/user-attachments/assets/099a1bf6-09ad-4a7d-bcae-d450dc44ee6d)

memory dump 

![image](https://github.com/user-attachments/assets/fed8889b-0173-43ab-af0c-fef6a2a9e905)

its simple api convert odata to sql query to EF.
Models its a database tables, complex and huge.

**Comment by julealgon** [comment]
> its simple api convert odata to sql query to EF.

Do you think you could share a small repro project that showcases this problem?

**Comment by Bosu123l** [comment]
sure! but I need some time to prepare example project.


________________________________
Od: Juliano Leal Goncalves ***@***.***>
Wysłane: piątek, 6 grudnia 2024 16:36
Do: OData/AspNetCoreOData ***@***.***>
DW: Łukasz Gawin ***@***.***>; Author ***@***.***>
Temat: Re: [OData/AspNetCoreOData] Memory leak Odata 8.2.7 (Issue #1369)


its simple api convert odata to sql query to EF.

Do you think you could share a small repro project that showcases this problem?

**Comment by Bosu123l** [comment]
Attached is the demo solution. When you run it, everything should set up and launch with Swagger. In Swagger, you can use OData elements, e.g., in the filter, you write "id eq 50", and in expand, you write "userWorkDetailUsers".

As you’ll see, the database consists of two models:

Users – which is the foundation of everything.
userWorkDetails – a dependent object that contains Users.
In the console, I enabled query logging, so you can observe how elegantly the OData query is converted into SQL using EF.

After around 7k requests, it can show significant memory usage, and it will remain so, as I described in the issue.

One more thing: the RolesAuthorizationHandler class is normally included in the middleware, but we’re not setting up AAD organization here. This class is responsible for verifying whether someone has appropriate permissions for an object, and it processes the EdmModel. This is my second suspicion regarding the issue.

Thanks for taking the time!
[demo.zip](https://github.com/user-attachments/files/18050079/demo.zip)

**Comment by corranrogue9** [comment]
@lukaszgawin in the diagrams that you have provided, when the memory drops back down, is that the result of a machine reset? They appear to be ~1 week apart each time.

Let's investigate the repro provided regardless.

**Comment by lukaszgawin** [comment]
Yes, the decrease in memory usage is due to a manual service reset. Intervals are every 5-6 days. We were responding to an outOfMemoryException

Wysłano z programu Outlook dla systemu Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by lukaszgawin** [comment]
The issue was observed after migrating to .NET 8. Initially, the service was running on .NET Core 3.1 and worked without any leaks, also without using the ODataModelBuilder. Over time, additional DB entities were added, along with the mentioned RolesAuthorizationHandler.

**Comment by habbes** [comment]
@lukaszgawin I'm curious about the `RolesAuthorizationHandler` as well. In the demo that you have shared, it's not referenced by any other class or method. At what point does it get invoked? This class seems suspicious because the `GetExpandProperties()` generates a model each time. Are these models properly disposed if they're not used again? If this method is called with the same type multiple times, should the model be cached and reused?

The `[Authorize]` attributes are commented out in the `UsersController`, is that intentional?

Lastly, I made a request in the demo and found some `EdmStructuralProperty` instances tied to `AuthorizationPolicyCache` instance in memory. However, I'm not yet sure how this connection is made. Do you know how the two could be related? Is it related to the `RolesAuthorizationHandler`?

![image](https://github.com/user-attachments/assets/eef951ce-e188-4a61-a5ed-9402e3bcc268)

In the screenshot that you took where you showed the OData objects using up memory, could you expand one of the nodes, e.g. `EdmStructuralProperty` so we can see what's keeping it in memory?

**Comment by lukaszgawin** [comment]
For the demo, I did not prepare all the scripts for the Microsoft tenant, including application Roles, assignments, etc.

Unfortunately, I cannot share the full application. It is integrated with Azure Active Directory, parses the token, and analyzes it based on roles to determine which entities you have access to. If a user includes an expand query, for example, for the Users entity, the **RolesAuthorizationHandler** is responsible for checking if the token contains the required role => refer to the dictionary in the class **Permissions** and thus either allows or denies access.

RolesAuthorizationHandler is registered as a scoped service in the object container, and in my opinion, such memory usage should not have such an impact. (Although it was initially a singleton, the result was the same.)

I created the dump as per your request, and here is the screenshot:

![image](https://github.com/user-attachments/assets/ec8edddd-7b39-4cdf-adc8-ce8202fd75d4)
Unfortunately, I do not see the AuthorizationPolicyCache you described in it.

**Comment by corranrogue9** [comment]
I got a repro in [this](https://github.com/OData/AspNetCoreOData/tree/issues/1369) branch when adding the authorization back. Still investigating, just wanted to give an update.

**Comment by habbes** [comment]
@corranrogue9  thanks for sharing the repro. I was able to confirm my suspicions thanks to the repro. As alluded to before, the issue stems from the fact the `RolesAuthorizationHandler` builds a new `IEdmModel` instance each time it's called, which is presumably on each request. @lukaszgawin it doesn't matter whether it's singleton or scoped service, if in each request it creates a new model.

The actual root cause is that ASP.NET Core OData caches the model in static cache (a concurrent dictionary). This is cache is used to store a mapping for CLR types to their corresponding EDM types for each model instance. This is by design, to make it quick to find the EdmType from the CLR type. Unfortunately such "magic" is not immediately obvious from working with IEdmModel in ASP.NET Core OData.

@lukaszgawin looking at `EdmStructuralProperty` in the memory profiler's table is a bit tricky since it's connected to so many other things. I would suggest searching "EdmModel` in the table and inspecting its instances and paths to root. You can see that the number of instances of the model grows proportionally with the number of requests. This is because an instance is created dynamically for each request, and the instances are never de-allocated since they're stored in a static cache. If you inspect the paths to root, you'll see one path leads to the `DefaultODataTypeMapper` which handles the mapping of CLR types to EDM types as explained above.

Here are some profiler snapshots after 1,500+ requests.

![image](https://github.com/user-attachments/assets/12408224-becb-4489-8aa7-946fccdf603b)

![image](https://github.com/user-attachments/assets/f8977911-3066-49da-a769-7e66c7987922)

![image](https://github.com/user-attachments/assets/2604b619-4ac8-49b4-a825-76ae9d279f66)

This is not an error in the library. Generally, you do not want to allocate a new model in each request. This particular scenario is easy to fix since you're creating a new model per parent type and you have limited set of types you want OData support for. You can easily cache the model per type in a dictionary and reuse the model the next time you see the same type. Here's a simple example:

```c#
// Create cache
private static readonly ConcurrentDictionary<Type, IEdmModel> TypeToModelCache = new();
```

And this is how you'd use it

```c#
var model = TypeToModelCache.GetOrAdd(parentType, _ =>
{
    var entityTypeConfiguration = builder.AddEntityType(parentType);
    builder.AddEntitySet(parentType.Name, entityTypeConfiguration);
    return builder.GetEdmModel();
});
```

I've extended @corranrogue9's repro with a fix in this branch: https://github.com/OData/AspNetCoreOData/compare/issues/1369-fix

Here's a profiler snapshot after 2,200+ requests, the number of models and model elements is now stable.

![image](https://github.com/user-attachments/assets/7bee25d2-8e12-46f3-93f1-ed8785715d36)

![image](https://github.com/user-attachments/assets/7ac4b703-b9aa-41a6-9f91-2637da663b8d)

**Comment by lukaszgawin** [comment]
@habbes 
> Unfortunately such "magic" is not immediately obvious from working with IEdmModel in ASP.NET Core OData.

true... :smile:

Thank you for your time. As soon as I get organized, I’ll start working on the fixes in RolesAuthorizationHandler, run the tests, and confirm.

Thank!

**Comment by julealgon** [comment]
> The actual root cause is that ASP.NET Core OData caches the model in static cache (a concurrent dictionary).

@habbes have you considered changing that design to rely on a singleton registration (or even `IMemoryCache`) instead of a private `static`? That would potentially give more control and visibility to people trying to do anything "different" when it comes to EDM model. For instance, they could inject your singleton abstraction and grab the EDM model from there (instead of recreating it) etc.

Depending on how OData exposes this, it could even open up the possibility of letting the consumer decide the lifetime of the "cache", or be able to create a different cache based on some criteria (API Versioning I imagine could also leverage that, potentially), while still leveraging the native "storage" that the library currently relies on.

Just a thought to maybe open up access to that cache so that consumers can also leverage it.

**Comment by lukaszgawin** [comment]
@habbes After applying the modification, it seems that the memory growth has stopped. Indeed, using such a cache isn't obvious at first glance.
I'll observe how it behaves over a longer period, but it's already clear that it no longer increases its memory usage.

Thank you for your time.

![image](https://github.com/user-attachments/assets/d9f7cbe4-10ce-42a0-8d17-eab3a77ccc15)

**Comment by lukaszgawin** [comment]
I confirm, it was an issue with that class. @habbes thanks for your help. Please consider suggestions of @julealgon.
